### ISSUE TITLE
"'node_modules/@prisma/client' has no exported member 'PrismaClient'." when using symlinks

### ISSUE THREAD

**Issue by amitdhiman000** [prisma/prisma#2584]
error TS2305: Module '"../../node_modules/@prisma/client"' has no exported member 'PrismaClient'.
import { PrismaClient } from '@prisma/client';

## How to reproduce
create a schema e.g.

```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:my-database.db"
}

model User {
  birthday  DateTime @default(now())
  createdAt DateTime @default(now())
  email     String   @unique
  gender    String
  id        Int      @default(autoincrement()) @id
  name      String
  password  String
  updatedAt DateTime @default(now())
}

```

1. Change DB name in prisma.schema to  `dev-database.db`
2. `npm run prisma:migrate`
3. `npm run prisma;generate`
all of these command are there in package.json script

```
"scripts": {
    "prisma:migrate": "prisma migrate save --schema=./app/data_access/prisma/schema.prisma --experimental && prisma migrate up --schema=./app/data_access/prisma/schema.prisma --experimental",
    "prisma:unmigrate": "prisma migrate down --schema=./app/data_access/prisma/schema.prisma --experimental",
    "prisma:introspect": "prisma introspect --schema=./app/data_access/prisma/schema.prisma",
    "prisma:generate": "prisma generate --schema=./app/data_access/prisma/schema.prisma"
}
```

It was still referring to old database name. and I am not able to run any queries from code.
I manually deleted  node_modules/.prisma  directory.

again followed steps 2 and 3
Now I started seeing this problem:

`error TS2305: Module '"../../node_modules/@prisma/client"' has no exported member 'PrismaClient'.
import { PrismaClient } from '@prisma/client';`

I checked the @prisma/client, .prisma directories present inside node_modules

## Expected behaviour
1.) after `prisma generate` new changes should automatically reflect in prisma client
2.) Should be able to import and use prisma client.

## Environment & setup

- OS:  Debian Ubuntu 20.04 LTS
- Database: SQLite

- Prisma version:
@prisma/cli          : 2.0.0-beta.6
Current platform     : debian-openssl-1.1.x

- Node.js version: 
v14.2.0

**Comment by pantharshit00** [comment]
Can you please make sure that you ran `prisma generate` after you deleted `.prisma` folder? Otherwise the code won't be there and the error is expected.


I followed the steps that you have described and I can't reproduce the problem:
![image](https://user-images.githubusercontent.com/22195362/83149888-48bf3900-a118-11ea-9dde-6fc5b1b39b93.png)


Otherwise, can you please share a reproduction in a git repo? That would be very helpful.

**Comment by richardwardza** [comment]
@amitdhiman000 - I'm not sure if it's a typo but in your "How to produce > step 3" you have `npm run prisma;generate` with a `;` and not a `:`

**Comment by amitdhiman000** [comment]
@pantharshit00 
I spent more time debugging the problem, doing some hit and trial

here is the analysis.

`./node_modules/.bin/prisma generate --schema=./app/data_access/prisma/schema.prisma`
OR
`npx prisma generate --schema=./app/data_access/prisma/schema.prisma`
this command seems to work.

But If I put same command in npm script 
```
scripts: {
"prisma:generate": "prisma generate --schema=./app/data_access/prisma/schema.prisma"
}
```

then execute it,
`npm run prisma:generate`
it is creating empty files in  node_modules/.prisma/client ,and then giving this error
`error TS2305: Module '"../../node_modules/@prisma/client"' has no exported member 'PrismaClient'`

so the problem boils down to.
**prisma generate doesn't work when it is triggered from npm scripts**

**Comment by amitdhiman000** [comment]
> @amitdhiman000 - I'm not sure if it's a typo but in your "How to produce > step 3" you have `npm run prisma;generate` with a `;` and not a `:`

that is a typo. please ignore it.
I have used it properly
`npm run prisma:generate`

**Comment by janpio** [comment]
Does it work if you call `npx prisma` in the script `prisma:generate`?

**Comment by amitdhiman000** [comment]
> Does it work if you call `npx prisma` in the script `prisma:generate`?

No, it doesn't work.

**Comment by pantharshit00** [comment]
I am still unable to reproduce it. Please share a reproduction inside of a repository.

**Comment by amitdhiman000** [comment]
> I am still unable to reproduce it. Please share a reproduction inside of a repository.

https://github.com/amitdhiman000/prisma_test

@pantharshit00 
There is one thing which I forgot to mention earlier.
**node_modules is a soft link to a node_modules in some other directoy.**
Actually this is just a hack, I am trying to share same node_modules accross many node projects.
just to avoid downloading simillar dependencies again.


I think this issue is very corner case, somehow I got trapped in this scenario.
I don't know how many people use hacks like this.

small recording:
https://drive.google.com/file/d/1I9QpZTjhIdb6fmtUnft0_NAFwnxO4YAF/view?usp=sharing

**Comment by TheBeachMaster** [comment]
Also having the  same issue 

```

**Comment by TheBeachMaster** [comment]
Also having the same issue

```
'PrismaClient' is not exported by node_modules/@prisma/client/index.js
1: import { PrismaClient } from '@prisma/client'
            ^
2: import cubejs from "@cubejs-client/core";
3: const jose = require("jose");
```

In my case, the `.prisma` folder exists in the node_modules directory and it's not empty

**Comment by timsuchanek** [comment]
Thanks for the reproduction @amitdhiman000! It's indeed an edge-case but we'll have a look.

@TheBeachMaster which package manager and which version of it are you using? (npm, yarn etc)

**Comment by TheBeachMaster** [comment]
@timsuchanek I'm using `yarn`. Having issue on both `node v12.12.0` and `node v 14.1.0`

**Comment by matheuschimelli** [comment]
I solvet this issue by updating my yarn to `1.22.4`, and im using `node 13.12.0`

**Comment by ssalbdivad** [comment]
Using [pnpm](https://pnpm.js.org/) & [rush](https://rushjs.io/), I got the same error.

It seems like the problem is an inconsistency in the way symlinks are handled during generation.

The ```.prisma``` directory ends up getting created next to wherever the symlink points:

```
✔ Generated Prisma Client to ./../../common/temp/node_modules/.pnpm/registry.npmjs.org/@prisma/client/2.0.1_@prisma+cli@2.0.1/node_modules/@prisma/client
```

whereas ```prisma``` ends up looking for it in the directory where the link itself exists:

```
Module '"../node_modules/@prisma/client"' has no exported member 'PrismaClient'.

2 import { PrismaClient } from "@prisma/client"
```

This can be resolved by specifying the generator output explicitly in your ```schema.prisma``` file:
```.prisma
generator client {
  provider      = "prisma-client-js"
  output = "../node_modules/.prisma/client"
}
```

**Comment by janpio** [comment]
I think we should turn your comment @ssalbdivad into a new issue "node_modules/@prisma/client"' has no exported member 'PrismaClient'. when using symlinks (for pnpm &rush)" or similar - then we can fix this explicit problem and close it. Can you copy your comment into an issue please?

**Comment by ssalbdivad** [comment]
I think what I encountered is very likely the same issue @amitdhiman000 was having. He explicitly mentioned:

> node_modules is a soft link to a node_modules in some other directoy.

If you'd still like me to create another issue, let me know and I will.

**Comment by janpio** [comment]
Ah, I missed that. I updated the issue title and will hide these comments in a bit @ssalbdivad

**Comment by HoverBaum** [comment]
Ran into this problem today trying to run the quickstart example on mac.

Solved it by upgrading yarn to 1.19.1 using `brew upgrade yarn`.

**Comment by pantharshit00** [comment]
@ssalbdivad Can you please try again with the latest version? We have removed a dependency which was relying on git path which pnpm doesn't seem to support so this might be already fixed.

**Comment by janpio** [comment]
Hey @ssalbdivad , could you have another look at this after @pantharshit00 could not reproduce this? With a newer Prisma version this should definitely not happen any more. If you can not reproduce this, we would like to close the issue.

**Comment by ssalbdivad** [comment]
@janpio Don't have the capacity to look at this right now, feel free to close 👍

**Comment by rafeedm** [comment]
> Using [pnpm](https://pnpm.js.org/) & [rush](https://rushjs.io/), I got the same error.
> 
> It seems like the problem is an inconsistency in the way symlinks are handled during generation.
> 
> The `.prisma` directory ends up getting created next to wherever the symlink points:
> 
> ```
> ✔ Generated Prisma Client to ./../../common/temp/node_modules/.pnpm/registry.npmjs.org/@prisma/client/2.0.1_@prisma+cli@2.0.1/node_modules/@prisma/client
> ```
> 
> whereas `prisma` ends up looking for it in the directory where the link itself exists:
> 
> ```
> Module '"../node_modules/@prisma/client"' has no exported member 'PrismaClient'.
> 
> 2 import { PrismaClient } from "@prisma/client"
> ```
> 
> This can be resolved by specifying the generator output explicitly in your `schema.prisma` file:
> 
> ```prisma
> generator client {
>   provider      = "prisma-client-js"
>   output = "../node_modules/.prisma/client"
> }
> ```

Tried this but getting 'ERR_MODULE_NOT_FOUND', any other fixes for this issue? I followed the quick start step by step

**Comment by johnfn** [comment]
@janpio as a heads up, I just ran into this today. Seems like the issue still exists.

I suspect it might be because I inadvertently used both yarn and npm on the same repo.

**Comment by janpio** [comment]
If you can somehow reproduce this, we are super happy to take another look @johnfn. Just let us know the steps and we'll jump on it.

**Comment by vjpr** [comment]
Ran into same issue. I had `prisma` installed globally in pnpm which caused the issue.

I had to install prisma locally, and then use `node_modules/.bin/prisma generate` to get it to work.

I'm not a fan of this approach - it's a bit of a hack.

Why don't we just generate the JS client in user's `src` dir? And then just require that the native client be installed as part of the `@prisma/client` dependency or another native dependency? Just like all other native deps (e.g. Cypress).

Is there anything in the native client that is "generated", or is it just copied from somewhere.

**Comment by un33k** [comment]
If you have installed `prisma` globally and have multiple generators and the last generator has the `output` specified explicitly, then  things will mess up.

**Example:**
```
generator client {
  provider = "prisma-client-js"
}

generator dbml {
  provider = "prisma-dbml-generator"
  output   = "./generated/dbml"
}```


**Solutions:**

A: Remove the global version of your `prisma`, then install it locally and reference the local in your packages.json as such. 

```
    "prisma:generate": "node ./node_modules/.bin/prisma generate --schema=<some-path>/prisma/schema.prisma"
```

Then run it as "yarn prisma:generate" or "npm run prisma:generate" and you should be fine.

Alternately: (B), remove all `output` fields from all generators if you have multiple. 

That should do it.

It would be great if we can leave  the`prisma` package to the history along with `prisma 1` and move forward, and just use the `@prisma/*` namespace.  Easy to avoid bugs like this and easier to debug issue.

**Comment by janpio** [comment]
Nice catch @un33k. Can you copy that into a new bug issue? Then we can reproduce and will work to fix this.

**Comment by Markkos89** [comment]
Just in case this saves someone... I was facing a simillar issue.. with lambdas and layers..

added this environment variable fixed everything and now my lambda con read from the layer :D

`NODE_PATH=./:/opt/nodejs/node_modules`

**Comment by maitrungduc1410** [comment]
> Using [pnpm](https://pnpm.js.org/) & [rush](https://rushjs.io/), I got the same error.
> 
> It seems like the problem is an inconsistency in the way symlinks are handled during generation.
> 
> The `.prisma` directory ends up getting created next to wherever the symlink points:
> 
> ```
> ✔ Generated Prisma Client to ./../../common/temp/node_modules/.pnpm/registry.npmjs.org/@prisma/client/2.0.1_@prisma+cli@2.0.1/node_modules/@prisma/client
> ```
> 
> whereas `prisma` ends up looking for it in the directory where the link itself exists:
> 
> ```
> Module '"../node_modules/@prisma/client"' has no exported member 'PrismaClient'.
> 
> 2 import { PrismaClient } from "@prisma/client"
> ```
> 
> This can be resolved by specifying the generator output explicitly in your `schema.prisma` file:
> 
> ```prisma
> generator client {
>   provider      = "prisma-client-js"
>   output = "../node_modules/.prisma/client"
> }
> ```

regarding this solution, I tried few times, it kept on failing, but then the last time was successful, not sure why, this is unstable somehow.

Here's [the PR they're working on to solve this issue](https://github.com/prisma/prisma/pull/4920), still haven't merged

**Comment by SamuelAure** [comment]
This worked for me https://stackoverflow.com/a/74273635/17995887

**Comment by nathan-mersha** [comment]
```
# Create app directory
WORKDIR /app

# A wildcard is used to ensure both package.json AND package-lock.json are copied
COPY package*.json ./
COPY prisma ./prisma/

# Install app dependencies
RUN npm install
```
copy the prisma directory just before running npm install, it fixed it for me

**Comment by amarjeet987** [comment]
I faced the same issue while cloning a repo to a new device. This worked for me :

1. `npm install` to install the deps
2. `npx prisma generate`
3. Error was still there. Closed VSCode and opened it again, it went away. I think it only looks for the generated files in step 2 when it's started.

**Comment by amarjeet987** [comment]
I faced the same issue while cloning a repo to a new device from github. I was working fine in my old device. This worked for me :

1. `npm install` to install the deps
2. `npx prisma generate`
3. Error was still there. Closed VSCode and opened it again, it went away. I think it only looks for the generated files in step 2 when it's started.

**Comment by omgismartinez** [comment]
You can try `restart TS server` on **vscode**

1. Open **Command palette**: Windows `Ctrl + Shift + P`
2. Type `TypeScript: Restart TS Server`
3. Enter and wait

**Comment by rristoferbowers** [comment]
> You can try `restart TS server` on **vscode**
> 
> 1. Open **Command palette**: Windows `Ctrl + Shift + P`
> 2. Type `TypeScript: Restart TS Server`
> 3. Enter and wait

LOL

**Comment by fulopdaniel** [comment]
A bit embarassing but maybe it will help someone. In my case I added some custom paths to `tsconfig.json` that began with `@` which messed things up when I was trying to import from `@prisma/client`.

**Comment by andreevpix** [comment]
> A bit embarassing but maybe it will help someone. In my case I added some custom paths to `tsconfig.json` that began with `@` which messed things up when I was trying to import from `@prisma/client`.

It's works!!!! Fine!!! O my God! Thanks!

**Comment by kylejw2** [comment]
> ```
> # Create app directory
> WORKDIR /app
> 
> # A wildcard is used to ensure both package.json AND package-lock.json are copied
> COPY package*.json ./
> COPY prisma ./prisma/
> 
> # Install app dependencies
> RUN npm install
> ```
> 
> copy the prisma directory just before running npm install, it fixed it for me

In case anyone's run into this issue in their Dockerfile, @nathan-mersha 's answer helped me

**Comment by suchcodemuchwow** [comment]
I had similar but different type of problem which might help some people. I was able to import the generated files but **blindly** which I mean autocomplete was not working on _Webstorm_. I figured out I earlier used **Mark Directory as** -> **Excluded**. Just another thing you might want to consider if you are Webstorm user. And don't forget to **Restart Typescript Service**

**Comment by tvankith2neoito** [comment]
I was setting up Lerna and I tried many solutions. [This](https://github.com/prisma/prisma/issues/2584#issuecomment-1291762470) one worked for me:

**Comment by mohammadrz003** [comment]
> ```
> # Create app directory
> WORKDIR /app
> 
> # A wildcard is used to ensure both package.json AND package-lock.json are copied
> COPY package*.json ./
> COPY prisma ./prisma/
> 
> # Install app dependencies
> RUN npm install
> ```
> 
> copy the prisma directory just before running npm install, it fixed it for me

I added `RUN npx prisma generate` after `RUN npm install` on top of your answer, and it worked for me!
so, it'll be like this at the end:
```
WORKDIR /usr/src/app

COPY package*.json ./

# New line added
COPY prisma ./prisma/

RUN npm install

# New line added
RUN npx prisma generate
```

**Comment by MariuzM** [comment]
I'm using PNPM so this `pnpm exec prisma generate` worked for me and i have my app deployed on Railway to the build script looks like `"build": "pnpm exec prisma generate && tsc -build",`

**Comment by marinawl** [comment]
Resolve by adding 'database' to next.config.js

`module.exports = {
  reactStrictMode: true,
  transpilePackages: ['ui', 'database'],
}`

**Comment by gllrmzndm** [comment]
Same issue, had to update Prisma.

```
───────────────────────────────────────────────────────┐
│  Update available 5.4.2 -> 5.5.2                        │
│  Run the following to update                            │
│    npm i --save-dev prisma@latest                       │
│    npm i @prisma/client@latest
```

**Comment by jarrodmedrano** [comment]
I had this error
` error TS2305: Module '"@prisma/client"' has no exported member 'PrismaClient'`
I was running a build in Docker.
This error went away after I did 
`docker system prune`
and
`docker volume prune -a`
Seems like I was just out of space

**Comment by marcelino-borges** [comment]
Just in case someone finds this to solve the problem as I did.
I had run the `npx prisma generate` and everything. I noticed that in `node_modules/.prisma/client/index.d.ts` prisma had created my types from my schema models, but I was importing the type like this:
```typescript
import { Prisma } from "@prisma/client";

export const createTenant = (tenant: Prisma.MyUserType) => {};
```

And this was giving me the error that prisma had not exported member with the name "MyUserType".
Everywhere on the internet (including Prisma docs) shows this approach.

So I noticed that changing my code to this made it work:
```typescript
import { MyUserType } from "@prisma/client";

export const createTenant = (tenant: MyUserType) => {};
```

Hope it helps someone!

**Comment by kesha100** [comment]
> Same issue, had to update Prisma. After installation.
> 
> ```
> ───────────────────────────────────────────────────────┐
> │  Update available 5.4.2 -> 5.5.2                        │
> │  Run the following to update                            │
> │    npm i --save-dev prisma@latest                       │
> │    npm i @prisma/client@latest
> ```
> 
> But I also did add:
> 
> ```
> // schema.prisma
> 
> generator client {
>   provider      = "prisma-client-js"
>   output = "../node_modules/.prisma/client"
> }
> ```

thank you! it helped me so much!

**Comment by zkdeve** [comment]
Property 'user' does not exist on type 'PrismaClient<PrismaClientOptions, never, DefaultArgs>'


**[ERROR]** 02:50:35 ⨯ Unable to compile TypeScript:
src/controllers/createUser.ts(14,39): error TS2339: Property 'user' does not exist on type 'PrismaClient<PrismaClientOptions, never, DefaultArgs>'.
src/controllers/createUser.ts(25,38): error TS2339: Property 'user' does not exist on type 'PrismaClient<PrismaClientOptions, never, DefaultArgs>'


`
model User {
  id          String   @id @default(cuid())
  authUserId  String   @unique
  name        String
  email       String
  phoneNumber String?
  address     String   @default("Gazipur, Dhaka")
  city        String   @default("Dhaka")
  state       String   @default("Dhaka")
  country     String   @default("Bangladesh")
  postalCode  String   @default("1000")
  isActive    Boolean  @default(true)
  isDeleted   Boolean  @default(false)
  createdAt   DateTime @default(now())
  updatedAt   DateTime @updatedAt

  @@index([authUserId])
}`

**Comment by teddis** [comment]
No luck for me today resolving prisma client models for my Typescript React/Server project. Has been working reliably for months. Using Prisma 5.13.0, Yarn 1.22.19, Monorepo (with symlinks), PyCharm IDE.

I started seeing problems only after I upgraded to 5.14.0. So I reverted and still having trouble. 

Running `yarn generate` => yarn script `npx prisma generate` has worked solidly 100% before today.

Prisma claims to build in expected location which is root node_modules folder which is symlinked across the monorepo. No dependency referencing @prisma/client will resolve my schema models.

✔ Generated Prisma Client (v5.13.0) to ./../../node_modules/.prisma/client in 231ms

However, deploying to Vercel builds just fine there. Same project, not sure how my local tree could be so different. 

I have cleaned node_modules recursively, `yarn cache clean`, invalidate IDE caches and restarted, restarted Typescript server. Turbo build fails, jest tests fail, nothing resolves.

**Comment by codex78** [comment]
> ```
> # Create app directory
> WORKDIR /app
> 
> # A wildcard is used to ensure both package.json AND package-lock.json are copied
> COPY package*.json ./
> COPY prisma ./prisma/
> 
> # Install app dependencies
> RUN npm install
> ```
> 
> copy the prisma directory just before running npm install, it fixed it for me

The line: `COPY prisma ./prisma` has solved my problem.

Thank you very much 🥇

**Comment by programming-with-ia** [comment]
i fix that issue only reload window in vscode
`ctrl+shift+p` then type `reload window` then hit `enter`
😆
